### PR TITLE
fix: Apply reporting timezones for BigQuery functions (#25080)

### LINKED ISSUE #12380: BigQuery filter by "current day" (today) showing items from yesterday
**Describe the bug**
If I set my filter to to be "today" (Today is the 22nd), it will show results from yesterday:
image

You can see in the image above that its set for "today" (today is the 22nd) and that it is showing items from the 21st (yesterday).

If I were to filter on April 22nd, it shows the correct results:
image

**To Reproduce**
1. Create a new question
2. Use a  BigQuery database
3. filter by a column that has dates, filter by "current" "day"
4. See its displaying stuff from yesterday.

**Expected behavior**
I expect it to show only items with the current day (April 22nd) not items in the 24 hour period.

**Screenshots**

**Information about your Metabase Installation:**
You can get this information by going to Admin -> Troubleshooting.

- Your browser and the version: (e.x. Chrome 52.1, Firefox 48.0, Safari 11.1, …)
- Your operating system: Windows 10, Chrome
- Your databases: Big Query
- Metabase version:v 0.35.3
- Metabase hosting environment: Docker
- Metabase internal database: MySQL

**Severity**
This is annoying, causes users to not trust the data.

### LINKED ISSUE #25080: Incorrect last-X-weeks/months/years calculation on BigQuery
**Describe the bug**
In b9cedccc7f8, @camsaul mostly fixed the BigQuery driver so that it could express the date calculations needed to drive X-weeks/months/years-ago calculations. This gets used whenever we create a filter by last X weeks/month/years.

However, this does not work correctly when the reporting timezone is not UTC. We can see why by testing out the parts of the generated SQL. For a "last 2 weeks, grouped by week" question, with the reporting timezone set to 'Pacific/Auckland' and the first day of the week set to Monday, the SQL looks sometime like this:

```
SELECT timestamp_trunc(`bi.mytable`.`event_at`, week(monday)) AS `event_at`, sum(`bi.mytable`.`somefield`) AS `sum`
FROM `bi.mytable`
WHERE (`bi.mytable`.`event_at` >= CAST(datetime_trunc(datetime_add(current_datetime(), INTERVAL -2 week), week(monday)) AS timestamp)
   AND `bi.mytable`.`event_at` < CAST(datetime_trunc(current_datetime(), week(monday)) AS timestamp))
GROUP BY `event_at`
ORDER BY `event_at` ASC
```

Note that the reporting timezone doesn't appear anywhere in the query. Running the expressions in the BigQuery console:

```
SELECT CAST(datetime_trunc(datetime_add(current_datetime(), INTERVAL -2 week), week(monday)) AS timestamp)
```

Produces `2022-08-15 00:00:00 UTC` - so the week cutoff is for UTC, not the local timezone.

The first issue is that the grouping part, `timestamp_trunc(`bi.mytable`.`event_at`, week(monday))`, needs to specify the timezone - `timestamp_trunc(`bi.mytable`.`event_at`, week(monday), 'Pacific/Auckland')` in my case.

The second part, where the relative calculation is performed, is more complicated. Firstly we would need to specify the timezone to `current_datetime()`, ie. `current_datetime('Pacific/Auckland')`. Unfortunately this isn't enough - the timezone still gets dropped, and the timestamp ends up at 00:00:00 UTC rather than 00:00:00 Pacific/Auckland. (I think BigQuery's handling of datetime timezones is poor here.)

Given that `timestamp_trunc` works correctly (if we give it the timezone), one option is to move the cast so that we do only the absolute minimum in converted datetimes, which is the relative time calculation - ie. just the `datetime_add` - then cast, then use `timestamp_trunc` with reporting timezone. This looks to me to be much simpler than trying to fix the timezone on the `datetime_add`/`datetime_trunc`.

In this approach, we don't even need to specify the timezone to `current_datetime` - it's fine to use UTC for the current time and the relative date calculation, as if we then convert it to a timestamp it's UTC anyway, and only then do we run the truncation in the specific reporting timezone.

So my query above would become:

```
SELECT timestamp_trunc(`bi.mytable`.`event_at`, week(monday), 'Pacific/Auckland') AS `event_at`, sum(`bi.mytable`.`somefield`) AS `sum`
FROM `bi.mytable`
WHERE (`bi.mytable`.`event_at` >= timestamp_trunc(CAST(datetime_add(current_datetime(), INTERVAL -2 week) AS timestamp), week(monday), 'Pacific/Auckland')
   AND `bi.mytable`.`event_at` < timestamp_trunc(CAST(current_datetime() AS timestamp), week(monday), 'Pacific/Auckland'))
GROUP BY `event_at`
ORDER BY `event_at` ASC
```

**To Reproduce**
1. Setup up a BigQuery table with a timestamp column
2. Set the reporting timezone to something other than UTC
3. Create a new question summarising some value by the timestamp at week granularity
4. Filter to the last 2 weeks
5. Observe that rows are filtered and grouped based on the UTC week boundaries, not the reporting timezone

**Expected behavior**
Week/month/year grouping should be performed in the reporting timezone for BigQuery.

**Information about your Metabase Installation:**

You can get this information by going to Admin -> Troubleshooting.

Your browser and the version: Chrome 104
Your operating system: macOS 12.3.1
Your databases: BigQuery
Metabase version: v0.44.1
Metabase hosting environment: Docker
Metabase internal database: H2

**Severity**
How severe an issue is this bug to you? Is this annoying, blocking some users, blocking an upgrade or blocking your usage of Metabase entirely?
Blocking use for financials and causing some weird UX behavior.

### PR DESCRIPTION
**Pull request by willbryant**
Fixes #25080 
Fixes #12380
Implements #10772

Flag bigquery as supporting reporting timezones.

Apply the chosen reporting timezone in:
* current_time(), current_date(), and current_datetime() calls.
* timestamp_trunc() calls.
* coercion from datetime to timestamp; to do this, we need to use the timestamp() function rather than a plain CAST statement (annoyingly, CAST to TIMESTAMP supports a time zone argument, but only for strings).
* extract() calls.

This makes these functions essentially match the behavior of other databases which use SET TIMEZONE instead.

After some false starts, the final implementation here tries not to change the overall way we deal with intervals any more than necessary, as the existing coed proved to have some complexity around when it should default to timestamps, when it needs to be cast to datetimes (since BigQuery only implements constant-interval operations on timestamps themselves), and how to cast between them.

The AtTimeZone record idea is borrowed from the presto_jdbc implementation, and in keeping with that, we've used hx/literal on the other places we paste the timezone in.

I've removed the test named timezones-test that used to check timezone support wasn't implemented; the real timezones_test.clj now covers bigquery, including result-rows-test.

###### Before submitting the PR, please make sure you do the following

- [ ] If you're attempting to fix a translation issue, please submit your changes to our [POEditor project](https://poeditor.com/join/project/ynjQmwSsGh) instead of opening a PR.

### Tests

- [x] Run the frontend and Cypress end-to-end tests with `yarn lint && yarn test`)
- [x] If there are changes to the backend codebase, run the backend tests with `clojure -X:dev:test`
- [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
      (unless it's a tiny documentation change).

### CODE REVIEW

**Comment by willbryant** [comment]
@flamber @salsakran https://www.metabase.com/docs/latest/developers-guide/code-reviews says that I should assign reviewers, but I don't seem to have the power to do that? Is that only for people in the org?

I think it should possibly be @camsaul @metamben as I believe they last worked on the code I've worked on here.

**Comment by flamber** [comment]
Hi @willbryant Thank you for the the PR. I'm guessing that this is kinda implementing #10772 and likely fixing #12380 ?
Wonder if it will have negative impact for some existing setups that might be using JVM timezone.
And yes, only orgs people can assign reviews.

**Comment by willbryant** [comment]
@flamber Thanks for fishing those out, yes it should resolve both.

My understanding is that this will bring BigQuery into line with the other databases so it should have the same JVM timezone behavior after. Definitely a behavior change, but hopefully a welcome one; setting the JVM timezone didn't seem to do anything to work around these issues without this patch.

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj [hunk 1]
```diff
@@ -345,6 +351,15 @@
 (def ^:private valid-time-extract-units
   #{:microsecond :millisecond :second :minute :hour})
 
+(defrecord AtTimeZone
+  ;; record type to support applying Presto's `AT TIME ZONE` operator to an expression
```
Comment (camsaul) [inline modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj:355]:
```suggestion
  ;; record type to support applying BigQuery's `AT TIME ZONE` operator to an expression
```

**Comment by camsaul** [comment]
Hi @willbryant. I've taken a look at this and generally I think this approach is solid. At any rate it's a real improvement over our current code. We've talked about doing something like this generally and dropping the whole `SET TIMEZONE` thing we do entirely. The work you've done here serves nicely as a proof-of-concept for doing that in the future for other databases we support.

As I'm sure you've noticed Our CI set up does not currently work for 3rd-party PRs since they require injection of sensitive credentials (i.e, the BigQuery cloud creds) and we don't want people to be able to write a PR that exfiltrates that stuff without our knowledge... I've eyeballed your PR and it looks okay to me in that regard. I've opened up #25585 to run the complete test suite against that PR.

If everything passes then I think we can go ahead and **merge** this PR. If some stuff fails then you'll need to fix it here and I can update #25585 with your changes and we can repeat the process.

I'd like to get @metamben's eyes on this stuff as well since he has been working on a lot of this stuff recently.

**Review by camsaul — APPROVED** [review]
Going to approve this for now, I'll try to remember to follow up after https://github.com/metabase/metabase/pull/25585 is done but if I forget then please ping me here so I remember if it's green. If it's red then fix it and ping me 

Still want @metamben's eyes on it tho.

**Comment by willbryant** [comment]
Thanks @camsaul.

Looks like I need to figure out how to make the main E2E tests run with a particular driver. 4 failures to figure out.

```
                                       clojure.lang.ExceptionInfo: Error executing query: com.google.cloud.bigquery.BigQueryException: Table "users" must be qualified with a dataset (e.g. dataset.table).
    parameters: [#t "2014-08-02T00:00-07:00[America/Los_Angeles]" #t "2014-08-03T00:00-07:00[America/Los_Angeles]"]
           sql: "-- Metabase\nSELECT `users`.`id`, `users`.`name`, `users`.`last_login` FROM `users` WHERE timestamp_trunc(`v3_test_data_with_timezones.users`.`last_login`, day, 'America/Los_Angeles') BETWEEN ? AND ? ORDER BY `users`.`id` ASC"
```

This is the wrong identifier quoting syntax for BigQuery. I don't think my branch has regressed this, but these tests didn't run before, so it didn't matter.

```
clojure.lang.ExceptionInfo: Error in with-temporary-setting-values: No test data type mapping for driver :bigquery-cloud-sdk for base type :type/TimeWithTZ; add an impl for field-base-type->sql-type.
    location: ********.driver/report-timezone
     setting: "report-timezone"
       value: nil
       java.lang.Exception: No test data type mapping for driver :bigquery-cloud-sdk for base type :type/TimeWithTZ; add an impl for field-base-type->sql-type.
```

This is new. BigQuery doesn't have an equivalent type, but neither do all the other DBs, so I assume I can work around this somehow.

```
FAIL in ********.driver.sql.parameters.substitute-test/e2e-parse-native-dates-test (substitute_test.clj:614)

:bigquery-cloud-sdk
Native dates should be parsed with the report timezone
expected: ["2018-04-18T00:00:00-07:00"]
  actual: (["2018-04-17T17:00:00-07:00"])
```

This looks like it might be real, as does:

```
FAIL in ********.query-processor.middleware.format-rows-test/format-rows-test (format_rows_test.clj:41)
```

**Review by metamben — APPROVED** [review]
Nice work! 🏅

File: modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk/query_processor_test.clj [hunk 2]
```diff
@@ -391,6 +391,25 @@
                    (sql.qp/format-honeysql :bigquery-cloud-sdk
                                            {:where (sql.qp/->honeysql :bigquery-cloud-sdk reconciled-clause)}))))))))
 
+  (testing "relative-datetime clauses on their own when a reporting timezone is set"
+    (doseq [timezone ["UTC" "US/Pacific"]]
+      (mt/with-temporary-setting-values [report-timezone timezone]
+      (doseq [[t [unit expected-sql]]
```
[the remark this answers is not in the document]
Comment (willbryant) [reply modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk/query_processor_test.clj:397]: Good spotting, fixed :)

**Comment by willbryant** [comment]
@camsaul Could you please update #25585 so we can check the E2E run?

I took the cheap way out and flagged BQ the same way(s) as other databases. Some of the tests look specific to the implementation of other DBs.

**Comment by camsaul** [comment]
updated #25585

**Comment by camsaul** [comment]
Looks like there are still a few failing tests:

https://circleci.com/gh/metabase/metabase/2154816?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

It might be a faster feedback loop if you run the tests locally... you just need to set `MB_BIGQUERY_CLOUD_SDK_TEST_PROJECT_ID` and `MB_BIGQUERY_CLOUD_SDK_TEST_SERVICE_ACCOUNT_JSON` to your project ID and a serialized JSON string (the same one you use to connect to BigQuery in Metabase) and then run the tests with 

```
DRIVERS=bigquery clojure -X:dev:drivers:drivers-dev:test
```

**Comment by willbryant** [comment]
@camsaul Thanks for those tips! I have been trying to run those tests, but it doesn't finish the run, as it terminates partway through with a fatal error due to the dashes in the `airports` dataset field names.

I assume if that this is not happening in CI because the dataset in your BigQuery project already has the dataset and it never tries to re-create it.

Do you remember creating the dataset manually to workaround a problem here?

If not, feel like dropping it in your test database so you can see what I see? :)

The error is:

```
Creating dataset "v3_airports"...
Deleted BigQuery dataset `delivereasy-data-test.v3_airports`.
Created BigQuery dataset `delivereasy-data-test.v3_airports`.
Failed to load BigQuery dataset "v3_airports".
{:via
 [{:type clojure.lang.ExceptionInfo,
   :message
   "Input to create-table! does not match schema: \n\n\t   [nil nil (named {(not (re-find #\"^[A-Za-z_]\\w{0,127}$\" \"iso-code\")) invalid-key} field-name->type)]  \n\n",
   :data
   {:type :schema.core/error,
    :schema
    [#schema.core.One{:schema (constrained Str "Non-blank string"), :optional? false, :name dataset-id}
     #schema.core.One{:schema (constrained Str "Non-blank string"), :optional? false, :name table-id}
     #schema.core.One{:schema
                      {#"^[A-Za-z_]\w{0,127}$"
                       (enum :TIME :BOOLEAN :DATE :NUMERIC :INTEGER :RECORD :FLOAT :STRING :DATETIME :TIMESTAMP)},
                      :optional? false,
                      :name field-name->type}],
    :value ["v3_airports" "continent" {"id" :INTEGER, "name" :STRING, "iso-code" :STRING}],
    :error [nil nil (named {(not (re-find #"^[A-Za-z_]\w{0,127}$" "iso-code")) invalid-key} field-name->type)]},
   :at
   [metabase.test.data.bigquery_cloud_sdk$eval213342$create_table_BANG___213347 invoke "bigquery_cloud_sdk.clj" 126]}],
 :trace
 [[metabase.test.data.bigquery_cloud_sdk$eval213342$create_table_BANG___213347 invoke "bigquery_cloud_sdk.clj" 126]
  [metabase.test.data.bigquery_cloud_sdk$load_tabledef_BANG_ invokeStatic "bigquery_cloud_sdk.clj" 286]
  [metabase.test.data.bigquery_cloud_sdk$load_tabledef_BANG_ invoke "bigquery_cloud_sdk.clj" 284]
  [metabase.test.data.bigquery_cloud_sdk$eval213549$fn__213551$fn__213564 invoke "bigquery_cloud_sdk.clj" 355]
  [metabase.util$do_with_auto_retries invokeStatic "util.clj" 463]
  [metabase.util$do_with_auto_retries invoke "util.clj" 450]
  [metabase.test.data.bigquery_cloud_sdk$eval213549$fn__213551 doInvoke "bigquery_cloud_sdk.clj" 344]
  [clojure.lang.RestFn invoke "RestFn.java" 425]
  [clojure.lang.MultiFn invoke "MultiFn.java" 234]
  [metabase.test.data.impl$create_database_BANG_$fn__83445$fn__83446 invoke "impl.clj" 98]
  [metabase.test.util.timezone$do_with_system_timezone_id invokeStatic "timezone.clj" 15]
  [metabase.test.util.timezone$do_with_system_timezone_id invoke "timezone.clj" 8]
  [metabase.test.data.impl$create_database_BANG_$fn__83445 invoke "impl.clj" 97]
  [clojure.core$binding_conveyor_fn$fn__5823 invoke "core.clj" 2047]
  [clojure.lang.AFn call "AFn.java" 18]
  [java.util.concurrent.FutureTask run "FutureTask.java" 264]
  [java.util.concurrent.ThreadPoolExecutor runWorker "ThreadPoolExecutor.java" 1136]
  [java.util.concurrent.ThreadPoolExecutor$Worker run "ThreadPoolExecutor.java" 635]
  [java.lang.Thread run "Thread.java" 833]],
 :cause
 "Input to create-table! does not match schema: \n\n\t   [nil nil (named {(not (re-find #\"^[A-Za-z_]\\w{0,127}$\" \"iso-code\")) invalid-key} field-name->type)]  \n\n",
 :data
 {:type :schema.core/error,
  :schema
  [#schema.core.One{:schema (constrained Str "Non-blank string"), :optional? false, :name dataset-id}
   #schema.core.One{:schema (constrained Str "Non-blank string"), :optional? false, :name table-id}
   #schema.core.One{:schema
                    {#"^[A-Za-z_]\w{0,127}$"
                     (enum :TIME :BOOLEAN :DATE :NUMERIC :INTEGER :RECORD :FLOAT :STRING :DATETIME :TIMESTAMP)},
                    :optional? false,
                    :name field-name->type}],
  :value ["v3_airports" "continent" {"id" :INTEGER, "name" :STRING, "iso-code" :STRING}],
  :error [nil nil (named {(not (re-find #"^[A-Za-z_]\w{0,127}$" "iso-code")) invalid-key} field-name->type)]}}
```

I have patched the regex to see if I can get past that, but now I'm just getting this:

```
Creating dataset "v3_airports"...
Deleted BigQuery dataset `delivereasy-data-test.v3_airports`.
Created BigQuery dataset `delivereasy-data-test.v3_airports`.
Deleted table `delivereasy-data-test.v3_airports.continent`
Created BigQuery table `delivereasy-data-test.v3_airports.continent`.
Inserting 6 rows like
{"name" "Africa", "iso-code" "AF", :id 1}

Inserting 6 rows like
{"name" "Africa", "iso-code" "AF", :id 1}

Inserting 6 rows like
{"name" "Africa", "iso-code" "AF", :id 1}

Inserting 6 rows like
{"name" "Africa", "iso-code" "AF", :id 1}

Inserting 6 rows like
{"name" "Africa", "iso-code" "AF", :id 1}

Inserting 6 rows like
{"name" "Africa", "iso-code" "AF", :id 1}

Failed to load BigQuery dataset "v3_airports".
{:via
 [{:type com.google.cloud.bigquery.BigQueryException,
   :message "Not found: Table delivereasy-data-test:v3_airports.continent",
   :at [com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc translate "HttpBigQueryRpc.java" 115]}
  {:type com.google.api.client.googleapis.json.GoogleJsonResponseException,
   :message
   "404 Not Found\nPOST https://www.googleapis.com/bigquery/v2/projects/delivereasy-data-test/datasets/v3_airports/tables/continent/insertAll?prettyPrint=false\n{\n  \"code\" : 404,\n  \"errors\" : [ {\n    \"domain\" : \"global\",\n    \"message\" : \"Not found: Table delivereasy-data-test:v3_airports.continent\",\n    \"reason\" : \"notFound\"\n  } ],\n  \"message\" : \"Not found: Table delivereasy-data-test:v3_airports.continent\",\n  \"status\" : \"NOT_FOUND\"\n}",
   :at
   [com.google.api.client.googleapis.json.GoogleJsonResponseException from "GoogleJsonResponseException.java" 146]}],
 :trace
 [[com.google.api.client.googleapis.json.GoogleJsonResponseException from "GoogleJsonResponseException.java" 146]
  [com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest
   newExceptionOnError
   "AbstractGoogleJsonClientRequest.java"
   118]
  [com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest
   newExceptionOnError
   "AbstractGoogleJsonClientRequest.java"
   37]
  [com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1
   interceptResponse
   "AbstractGoogleClientRequest.java"
   428]
  [com.google.api.client.http.HttpRequest execute "HttpRequest.java" 1111]
  [com.google.api.client.googleapis.services.AbstractGoogleClientRequest
   executeUnparsed
   "AbstractGoogleClientRequest.java"
   514]
  [com.google.api.client.googleapis.services.AbstractGoogleClientRequest
   executeUnparsed
   "AbstractGoogleClientRequest.java"
   455]
  [com.google.api.client.googleapis.services.AbstractGoogleClientRequest
   execute
   "AbstractGoogleClientRequest.java"
   565]
  [com.google.cloud.bigquery.spi.v2.HttpBigQueryRpc insertAll "HttpBigQueryRpc.java" 492]
  [com.google.cloud.bigquery.BigQueryImpl$28 call "BigQueryImpl.java" 1047]
  [com.google.cloud.bigquery.BigQueryImpl$28 call "BigQueryImpl.java" 1044]
  [com.google.api.gax.retrying.DirectRetryingExecutor submit "DirectRetryingExecutor.java" 105]
  [com.google.cloud.RetryHelper run "RetryHelper.java" 76]
  [com.google.cloud.RetryHelper runWithRetries "RetryHelper.java" 50]
  [com.google.cloud.bigquery.BigQueryImpl insertAll "BigQueryImpl.java" 1043]
  [metabase.test.data.bigquery_cloud_sdk$insert_data_BANG_ invokeStatic "bigquery_cloud_sdk.clj" 212]
  [metabase.test.data.bigquery_cloud_sdk$insert_data_BANG_ invoke "bigquery_cloud_sdk.clj" 204]
  [metabase.test.data.bigquery_cloud_sdk$load_tabledef_BANG_$fn__209538 invoke "bigquery_cloud_sdk.clj" 291]
  [metabase.test.data.bigquery_cloud_sdk$load_tabledef_BANG_ invokeStatic "bigquery_cloud_sdk.clj" 290]
  [metabase.test.data.bigquery_cloud_sdk$load_tabledef_BANG_ invoke "bigquery_cloud_sdk.clj" 284]
  [metabase.test.data.bigquery_cloud_sdk$eval209561$fn__209563$fn__209576 invoke "bigquery_cloud_sdk.clj" 355]
  [metabase.util$do_with_auto_retries invokeStatic "util.clj" 461]
  [metabase.util$do_with_auto_retries invoke "util.clj" 450]
  [metabase.util$do_with_auto_retries invokeStatic "util.clj" 468]
  [metabase.util$do_with_auto_retries invoke "util.clj" 450]
  [metabase.util$do_with_auto_retries invokeStatic "util.clj" 468]
  [metabase.util$do_with_auto_retries invoke "util.clj" 450]
  [metabase.test.data.bigquery_cloud_sdk$eval209561$fn__209563 doInvoke "bigquery_cloud_sdk.clj" 344]
  [clojure.lang.RestFn invoke "RestFn.java" 425]
  [clojure.lang.MultiFn invoke "MultiFn.java" 234]
  [metabase.test.data.impl$create_database_BANG_$fn__81091$fn__81092 invoke "impl.clj" 98]
  [metabase.test.util.timezone$do_with_system_timezone_id invokeStatic "timezone.clj" 15]
  [metabase.test.util.timezone$do_with_system_timezone_id invoke "timezone.clj" 8]
  [metabase.test.data.impl$create_database_BANG_$fn__81091 invoke "impl.clj" 97]
  [clojure.core$binding_conveyor_fn$fn__5823 invoke "core.clj" 2047]
  [clojure.lang.AFn call "AFn.java" 18]
  [java.util.concurrent.FutureTask run "FutureTask.java" 264]
  [java.util.concurrent.ThreadPoolExecutor runWorker "ThreadPoolExecutor.java" 1136]
  [java.util.concurrent.ThreadPoolExecutor$Worker run "ThreadPoolExecutor.java" 635]
  [java.lang.Thread run "Thread.java" 833]],
 :cause
 "404 Not Found\nPOST https://www.googleapis.com/bigquery/v2/projects/delivereasy-data-test/datasets/v3_airports/tables/continent/insertAll?prettyPrint=false\n{\n  \"code\" : 404,\n  \"errors\" : [ {\n    \"domain\" : \"global\",\n    \"message\" : \"Not found: Table delivereasy-data-test:v3_airports.continent\",\n    \"reason\" : \"notFound\"\n  } ],\n  \"message\" : \"Not found: Table delivereasy-data-test:v3_airports.continent\",\n  \"status\" : \"NOT_FOUND\"\n}"}

Deleted BigQuery dataset `delivereasy-data-test.v3_airports`.
```

**Comment by willbryant** [comment]
Should be green now. Created #25742 if you wanna merge the column name fixes first.

**Comment by sockpapi** [comment]
When will this be deployed to production?

**Review by camsaul — CHANGES_REQUESTED** [review]
Please don't change the test datasets, it will almost certainly end up breaking tests for other drivers

This should not be needed anyway because BigQuery implements `metabase.driver.ddl.interface/format-name` and converts identifiers to `snake_case` in that method (line 89 of `metabase.test.data.bigquery-cloud-sdk`)

**Comment by willbryant** [comment]
@camsaul Unfortunately that doesn’t happen, ValidFieldNames rejects the field name. 

I tried changing that regex to allow dashes but it just went ahead and tried to create the column named as is. (It mostly then ignored the table create failures due to some explicit ignores in the test setup code, which seem a bit too broad, so the eventual failures were pretty misleading.)

Definitely understand the desire not to alter test data in general but I believe this is the only dataset that used dashes and I think I have kept everything up to date with the change?

**Comment by camsaul** [comment]
@willbryant I'll try to remember to take a look at it tomorrow, if I don't post something here please ping me, I probably forgot 😄

**Comment by willbryant** [comment]
@camsaul Thanks! I can see a bunch of places in test/ where format-name will get applied to table names, but not to column names (unlike the main app code proper).

If it helps, I posted some of the errors above in https://github.com/metabase/metabase/pull/25451#issuecomment-1264606545

**Comment by nemanjaglumac** [comment]
@willbryant your changes have been merged in https://github.com/metabase/metabase/pull/25889.

Thank you so much for your contribution.
It's greatly appreciated.

Also, sorry it took so long to merge it.

**Pull request closed without merging**